File: 234. Palindrome Linked List/solution.ts
/** Definition for singly-linked list. */
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

function isPalindrome(head: ListNode | null): boolean {
    // 1. Find the midpoint using Floyd's Algorithm
    let slow: ListNode | null = head;
    let fast: ListNode | null = head;
    while (fast && fast.next) {
        slow = slow!.next;
        fast = fast.next.next;
    }

    // 2. Reverse the second half.
    let prev: ListNode | null = null;
    while (slow) {
        let next: ListNode | null = slow.next;
        slow.next = prev;
        prev = slow;
        slow = next;
    }

    // 3. Traverse from both ends and verify palindrome.
    let left: ListNode | null = head;
    let right: ListNode | null = prev;
    while (left && right) {
        if (left.val != right.val) {
            return false;
        }
        left = left.next;
        right = right.next;
    }

    return true;
};

File: 234. Palindrome Linked List/solution.py
from typing import *

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # Step 1: Find Midpoint
        slow, fast = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        # Step 2: Reverse right half
        prev = None
        while slow:
            nxt = slow.next
            slow.next = prev
            prev = slow
            slow = nxt
        
        # Step 3: Start from both ends, validate
        left, right = head, prev
        while left and right:
            if left.val != right.val:
                return False
            left = left.next
            right = right.next
        
        return True